lib/muser.h

```diff
@@ -314,85 +217,179 @@ typedef struct {
     struct lm_sparse_mmap_areas *mmap_areas;
 } lm_migration_t;
 
+/*
+ * Attaching to the transport is non-blocking. The library will not attempt
+ * to attach during context creation time. The caller must then manually
+ * call lm_ctx_try_attach(), which is non-blocking, as many times as
+ * necessary.
+ */
+#define LM_FLAG_ATTACH_NB  (1 << 0)
+
 /**
- * Device information structure, used to create the lm_ctx.
- * To be filled and passed to lm_ctx_create()
+ * Creates libmuser context.
+ * @path: path to socket file.
+ * @flags: context flag
+ * @log: log function
+ * @log_lvl: logging level
+ * @trans: transport type
+ * @pvt: private data
+ * @returns the lm_ctx to be used or NULL on error. Sets errno.
  */
-typedef struct {
-    char            *uuid;
+lm_ctx_t *lm_create_ctx(const char *path, int flags, lm_log_fn_t *log,
+                        lm_log_lvl_t log_lvl, lm_trans_t trans, void *pvt);
 
-    /*
-     * Private data passed to various lm_XXX functions.
-     */
-    void            *pvt;
+//TODO: Check other PCI header registers suitable to be filled by device.
+//      Or should we pass whole lm_pci_hdr_t to be filled by user.
+/**
+ * Setup PCI header data.
+ * @lm_ctx: the libmuser context
+ * @id: Device and vendor ID
+ * @ss: Subsystem vendor and device ID
+ * @cc: Class code
+ * @extended: support extented PCI config space
+ */
+int lm_setup_pci_hdr(lm_ctx_t *lm_ctx, lm_pci_hdr_id_t *id, lm_pci_hdr_ss_t *ss,
```

Comment:
1. Added "lm_pci" prefix to the PCI-specific functions.
2. We need to study the PCI header and check which all registers needs to be specified by user. Today we only accept id, ss and cc. If there are more then we can think of getting this whole standard/required header registers from user.  Yes but I agree we don't get the facility to set some registers by ourself.
3. IMO better to have less functions instead of having separate functions for each register.